core/execution/engine_snapshot.go

```diff
@@ -52,6 +55,105 @@ func (e *Engine) marketsStates() ([]*types.ExecMarket, []types.StateProvider) {
 	return mks, e.newGeneratedProviders
 }
 
+func (e *Engine) spotMarketsStates() ([]*types.ExecSpotMarket, []types.StateProvider) {
+	mkts := len(e.spotMarketsCpy)
+	if mkts == 0 {
+		return nil, nil
+	}
+	mks := make([]*types.ExecSpotMarket, 0, mkts)
+
+	// we don't really know how many new markets there are so don't bother with the calculation
+	e.newGeneratedProviders = []types.StateProvider{}
+	for _, m := range e.spotMarketsCpy {
+		e.log.Debug("serialising spot market", logging.String("id", m.Mkt().ID))
+		mks = append(mks, m.GetState())
+		if _, ok := e.generatedProviders[m.GetID()]; !ok {
+			e.newGeneratedProviders = append(e.newGeneratedProviders, m.GetNewStateProviders()...)
+			e.generatedProviders[m.GetID()] = struct{}{}
+		}
+	}
+
+	return mks, e.newGeneratedProviders
+}
+
+func (e *Engine) restoreSpotMarket(ctx context.Context, em *types.ExecSpotMarket) (*spot.Market, error) {
+	marketConfig := em.Market
+	if len(marketConfig.ID) == 0 {
+		return nil, ErrNoMarketID
+	}
+
+	// ensure the asset for this new market exists
+	asts, err := marketConfig.GetAssets()
+	if err != nil {
+		return nil, err
+	}
+
+	assetDetatils := []*assets.Asset{}
+	for _, asset := range asts {
+		if !e.collateral.AssetExists(asset) {
+			return nil, fmt.Errorf(
+				"unable to restore a spot market %q with an invalid %q asset",
+				marketConfig.ID,
+				asset,
+			)
+		}
+		ad, err := e.assets.Get(asset)
+		if err != nil {
+			e.log.Error("Failed to restore a market, unknown asset",
+				logging.MarketID(marketConfig.ID),
+				logging.String("asset-id", asset),
+				logging.Error(err),
+			)
+			return nil, err
+		}
+		assetDetatils = append(assetDetatils, ad)
+	}
+
+	nextMTM := time.Unix(0, em.NextMTM)
```

Comment:
any chance for nextMTM to be the zero time? That would be an issue to serialise it like this then.

Reply:
indeed.

Reply:
what do you mean?